---
title: "MixviR_Paper_Analyses"
author: "Mike Sovic"
date: "5/6/2022"
output: 
  html_document:
    toc: true
---

Below is code associated with analyses in the paper describing MixviR. Note that in some cases infiles will need to be downloaded and paths updated prior to running the code. 

# R Setup

```{r, warning=FALSE, message=FALSE}
library(MixviR)
library(tidyverse)
library(patchwork)
library(magrittr)
```

# SARS-CoV-2 Validation

## Generate VCF Files

### Raw Data

Raw data files (fastq; N=7 mixtures) are available from NCBI's short read archive (SRA), Bioproject accession PRJNA827817. 

### Alignment

Alignments of each file were performed with Illumina's Dragen Somatic app using the Wuhan reference to generate bams.

### Variant Calling (bcftools)

VCF files were generated with bcftools per the code below and added to dir "./covid_validation/vcfs/".

```{bash, eval = FALSE}
#for each bam file...
bcftools-1.11/bcftools mpileup -f <FASTA REFERENCE> -d 4000 -q 60 -Q 30 -L 4500 --ff UNMAP,SECONDARY,QCFAIL -a FORMAT/AD,FORMAT/ADF,FORMAT/ADR,FORMAT/DP,INFO/AD,INFO/ADF,INFO/ADR <INPUT BAM> | bcftools-1.11/bcftools call -m -A -Ov -o vcfs/<OUT>_temp.vcf

bcftools-1.11/bcftools norm vcfs/<OUT>_temp.vcf -c w -f <FASTA REFERENCE> -m -both -Ov -o <OUT>.vcf
```

## MixviR Analysis

### `call_mutations()`
```{r}
samp_muts <- call_mutations(sample.dir = "./covid_validation/vcfs/", 
                            reference = "Wuhan")

```

### `estimate_lineages()`

```{r}
lin.table <- estimate_lineages(muts.df = samp_muts,
                  lineage.muts = "https://raw.githubusercontent.com/mikesovic/MixviR/main/mutation_files/outbreak_20220217.csv",
                  outfile.name = "./covid_validation/estimated_proportions.csv")

```

## Generate Plots

### Target Mutation Proportions Plot (2A)

```{r, warning=FALSE, message=FALSE}

#combine results for number of mutations observed from "estimated_proportions.csv" with expected data in "SarsCov2_expected_target_mut_numbers.csv" 

covid_mut_prop_expected <- readr::read_csv("https://raw.githubusercontent.com/mikesovic/MixviR/main/mixvir_paper/SarsCov2_expected_target_mut_numbers.csv")

covid_mut_prop_observed <- readr::read_csv("./covid_validation/estimated_proportions.csv") %>%
  dplyr::rename("Value" = "Num_Target_Muts") %>%
  dplyr::mutate("Targets" = "Detected") %>%
  dplyr::select(Sample, Lineage, Targets, Value) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix1.vcf", "Mixture 1")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix2.vcf", "Mixture 2")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix3.vcf", "Mixture 3")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix4.vcf", "Mixture 4")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix5.vcf", "Mixture 5")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix6.vcf", "Mixture 6")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix7.vcf", "Mixture 7")) %>%
  dplyr::mutate("Lineage" = stringr::str_replace_all(Lineage, "Omicron_", ""))

covid_mut_prop_results <- dplyr::bind_rows(covid_mut_prop_expected, covid_mut_prop_observed) %>%
  dplyr::mutate("Targets" = factor(Targets, levels = c("Total", "Detected")))

covid_mut_prop_plot <- ggplot2::ggplot(covid_mut_prop_results, ggplot2::aes(x = Lineage, y = Value, fill = Targets)) +
  ggplot2::geom_bar(stat = "identity", position = "dodge") +
  ggplot2::scale_fill_brewer(palette = "Paired") +
  ggplot2::facet_wrap(ggplot2::vars(Sample), scales = "free") +
  ggplot2::theme(axis.title.x = ggplot2::element_blank()) +
  ggplot2::ggtitle("Target Mutations Identified From Mixtures of Sars-Cov-2 Lineages") +
  ggplot2::labs(y = "Number Of Target Mutations",
                title = "MixviR Validation: Sars-Cov-2 Mixtures",
                subtitle = "Lineage-Characteristic Mutations Detected") + 
  ggplot2::theme(legend.direction = "horizontal", 
                 legend.position = c(0.86, 0.08), 
                 legend.justification = c(1, 0),
                 legend.key.size = ggplot2::unit(0.7, 'cm'),
                 legend.title = ggplot2::element_blank(),
                 legend.text = ggplot2::element_text(size = 12),
                 strip.text = ggplot2::element_text(size = 12),
                 axis.text = ggplot2::element_text(size = 11),
                 title = ggplot2::element_text(size = 15))

ggplot2::ggsave(filename = "./Figure_2A.tiff", dpi = 600)
```

### Lineage Proportions Plot (2B)

```{r, warning=FALSE, message=FALSE}

#combine results for estimates of lineage proportions observed from "estimated_proportions.csv" with expected data in "SarsCov2_expected_lineage_proportions.csv" 

covid_prop_expected <- readr::read_csv("https://raw.githubusercontent.com/mikesovic/MixviR/main/mixvir_paper/SarsCov2_expected_lineage_proportions.csv")

covid_prop_observed <- readr::read_csv("./covid_validation/estimated_proportions.csv") %>%
  dplyr::rename("Scaled Proportion Mean" = "Mean_Freq") %>%
  dplyr::mutate("Type" = "MixviR Estimate") %>%
  dplyr::select(Sample, Lineage, Type, `Scaled Proportion Mean`) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix1.vcf", "Mixture 1")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix2.vcf", "Mixture 2")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix3.vcf", "Mixture 3")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix4.vcf", "Mixture 4")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix5.vcf", "Mixture 5")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix6.vcf", "Mixture 6")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "Mix7.vcf", "Mixture 7")) %>%
  dplyr::mutate("Lineage" = stringr::str_replace_all(Lineage, "Omicron_", ""))

covid_results_prop <- dplyr::bind_rows(covid_prop_expected, covid_prop_observed)

covid_lin_prop_plot <- ggplot2::ggplot(covid_results_prop, ggplot2::aes(x = Type, y = `Scaled Proportion Mean`, fill = Lineage)) +
  ggplot2::geom_bar(stat = "identity", position = "stack") +
  ggplot2::scale_fill_brewer(palette = "Paired") +
  ggplot2::facet_wrap(ggplot2::vars(Sample)) +
  ggplot2::theme(axis.title.x = ggplot2::element_blank()) +
  ggplot2::labs(y = "Proportion In Sample",
                title = "",
                subtitle = "Estimates Of Lineage Proportions") +
  ggplot2::theme(legend.direction = "horizontal", 
                 legend.position = c(0.95, 0.08), 
                legend.justification = c(1, 0),
                legend.key.size = ggplot2::unit(0.7, 'cm'),
                legend.title = ggplot2::element_blank(),
                legend.text = ggplot2::element_text(size = 12),
                strip.text = ggplot2::element_text(size = 12),
                axis.text = ggplot2::element_text(size = 11),
                title = ggplot2::element_text(size = 15)) +
  ggplot2::scale_x_discrete(labels= c("Expected", "MixviR\nEstimate"))

ggplot2::ggsave(filename = "./Figure_2B.tiff", dpi = 600)

```


### Combine SARS-CoV-2 Plots (Fig 2)

```{r, fig.width=10}
sars_plot <- covid_mut_prop_plot + plot_spacer() + covid_lin_prop_plot + 
  plot_layout(widths = c(4.4, 0.2 ,4.4))
sars_plot
grid::grid.text("B", x = 0.96, y = 0.06, gp = grid::gpar(fontsize = 20, family = "Times", fontface = "bold"))
grid::grid.text("A", x = 0.46, y = 0.06, gp = grid::gpar(fontsize = 20, family = "Times", fontface = "bold"))

ggplot2::ggsave(filename = "./Figure_2.tiff", dpi = 600)

```


# M. tuberculosis Validation

## Generate VCF Files

These are stored in dir "mtub_validation/vcfs/"

### Download fastqs

Data from SRA

```{bash, eval = FALSE}
#for each of "ERR221660, ERR221662, ERR221663, ERR221664, ERR221665, ERR221666"
module load sratoolkit
prefetch ${i}
fastq-dump --split-files ${i}.sra
```

### Download M. tuberculosis Reference

```{bash, eval = FALSE}
mkdir ref
curl -o ref/mtub.fa.gz https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/195/955/GCF_000195955.2_ASM19595v2/GCF_000195955.2_ASM19595v2_genomic.fna.gz
cd ref
gunzip mtub.fa.gz
```

### Align To Reference (Hisat2)

```{bash, eval = FALSE}
#build index
module load hisat2
hisat2-build mtub.fa mtub

#for each "ERR221660, ERR221662, ERR221663, ERR221664, ERR221665, ERR221666"
hisat2 -x ref/mtub -1 ${i}_1.fastq -2 ${i}_2.fastq -S ${i}.sam --un-conc-gz paired_unaligned_concord_${i}.fastq.gz --al-conc-gz paired_aligned_${i}.fastq.gz --summary-file summary_${i}.txt --rg-id ${i} --rg SM:${i}
```

### Mark PCR Duplicates

```{bash, eval = FALSE}
#for each of "ERR221660, ERR221662, ERR221663, ERR221664, ERR221665, ERR221666"
samtools sort -@ 4 -o ${i}.sorted.bam ${i}.sam
java -jar $PICARD MarkDuplicates I=${i}.sorted.bam O=picard_out/${i}.bam M=picard_out/metrics/${i}
```

### Variant Calling (GATK)
```{bash, eval = FALSE}
module load picard
java -jar $PICARD CreateSequenceDictionary R=ref/mtub.fa O=ref/mtub.dict

module load samtools
cd ref
samtools faidx mtub.fa
cd ../

#for each of "ERR221660, ERR221662, ERR221663, ERR221664, ERR221665, ERR221666"
java -jar $PICARD BuildBamIndex I=picard_out/${i}.bam
module load gatk/4.1.2.0
gatk HaplotypeCaller -ploidy 2 -I picard_out/${i}.bam -R ref/mtub.fa -O ${i}.vcf
gatk LeftAlignAndTrimVariants -V ${i}.vcf -R ref/mtub.fa --split-multi-allelics -O ${i}_split.vcf

```

## Create MixviR Bed File
```{bash, eval = FALSE}
cd ref
curl -L -o mtub.gff.gz https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/195/955/GCF_000195955.2_ASM19595v2/GCF_000195955.2_ASM19595v2_genomic.gff.gz

gunzip mtub.gff.gz

grep -v -E "^#" mtub.gff > temp1.txt
awk 'BEGIN {OFS="\t"} $3=="gene" {print $1,$4,$5,$6,$7,$9}' temp1.txt > genes.txt

sed -E 's/(ID=.+Name=)(.+)(;experiment.+)/\2/' genes.txt > names.txt
sed -E 's/(ID=.+Name=)(.+)(;gbkey.+)/\2/' names.txt > names2.txt

awk 'BEGIN {OFS="\t"} {print $1,$2,$3,$6,$4,$5}' names2.txt > mixvir_mtub.bed

```

## Get Characteristic Mutations For Target Lineages
Mutations associated with lineages designated "L1" and "L3" (corresponding to lineages EAI6-BGD1 and CAS1-Kili, respectively) were obtained from Table S3 of Merker et al (2020). These are available from "https://raw.githubusercontent.com/mikesovic/MixviR/main/mixvir_paper/lineage_muts_Mtub_EAI_CAS.csv".

## Clean Up Reference Chromosome Names

```{bash, eval = FALSE}
cd ref
sed -E 's/[[:space:]].*$//' mtub.fa > mtub_clean.fa

```
## MixviR Analysis

### `call_mutations()`
```{r, eval = FALSE}

# note this analysis is slow due to the size of the genome - ~5-10 min/sample.

samp_muts <- call_mutations(sample.dir = "./mtub_validation/vcfs/", 
               fasta.genome = "./mtub_validation/ref/mtub_clean.fa", 
               bed = "./mtub_validation/ref/mixvir_mtub.bed",
               genetic.code.num = "11",
               write.all.targets = TRUE, 
               lineage.muts = "https://raw.githubusercontent.com/mikesovic/MixviR/main/mixvir_paper/lineage_muts_Mtub_EAI_CAS.csv",
               write.mut.table = TRUE,
               outfile.name = "mtub_validation/Mtub_samp_mutations.tsv")

```

### `estimate_lineages()`

```{r}
lin.table <- estimate_lineages(lineage.muts = "https://raw.githubusercontent.com/mikesovic/MixviR/main/mixvir_paper/lineage_muts_Mtub_EAI_CAS.csv",
                  outfile.name = "mtub_validation/estimated_proportions.csv",
                  read.muts.from = "mtub_validation/Mtub_samp_mutations.tsv",
                  report.all = TRUE)

```

## Generate Plots

### Target Mutation Proportions Plot (3A)

```{r, warning=FALSE, message=FALSE}

#combine results for number of mutations observed from "estimated_proportions.csv" with expected data in "Mtub_expected_target_mut_numbers.csv" 

mtub_mut_num_expected <- readr::read_csv("https://raw.githubusercontent.com/mikesovic/MixviR/main/mixvir_paper/Mtub_expected_target_mut_numbers.csv")

mtub_mut_num_observed <- readr::read_csv("./mtub_validation/estimated_proportions.csv") %>%
  dplyr::rename("Value" = "Num_Target_Muts") %>%
  dplyr::mutate("Targets" = "Detected") %>%
  dplyr::select(Sample, Lineage, Targets, Value) %>%
  dplyr::mutate("Lineage" = stringr::str_replace_all(Lineage, "L1 \\(EAI\\)", "EAI")) %>%
  dplyr::mutate("Lineage" = stringr::str_replace_all(Lineage, "L3 \\(CAS\\)", "CAS")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "_split.vcf", ""))

mtub_results_mut_num <- dplyr::bind_rows(mtub_mut_num_expected, mtub_mut_num_observed) %>%
  dplyr::mutate("Targets" = factor(Targets, levels = c("Total", "Detected")))

mtub_mut_prop_plot <- ggplot2::ggplot(mtub_results_mut_num, ggplot2::aes(x = Lineage, y = Value, fill = Targets)) +
  ggplot2::geom_bar(stat = "identity", position = "dodge") +
  ggplot2::scale_fill_brewer(palette = "Paired") +
  ggplot2::facet_wrap(ggplot2::vars(Sample), scales = "free") +
  ggplot2::theme(axis.title.x = ggplot2::element_blank()) +
  ggplot2::ggtitle("Target Mutations Identified From Mixtures of M. tuberculosis Lineages") +
  ggplot2::labs(y = "Number Of Target Mutations",
                title = "MixviR Validation: M. tuberculosis Mixtures",
                subtitle = "Lineage-Characteristic Mutations Detected") + 
  ggplot2::theme(legend.title = ggplot2::element_blank(),
                 legend.text = ggplot2::element_text(size = 10),
                 strip.text = ggplot2::element_text(size = 12),
                 axis.text = ggplot2::element_text(size = 11),
                 title = ggplot2::element_text(size = 15))

ggplot2::ggsave(filename = "./Figure_3A.tiff", dpi = 600)

```

### Lineage Proportions Plot (3B)

```{r, warning=FALSE, message=FALSE}
#combine results for estimates of lineage proportions observed from "estimated_proportions.csv" with expected data in "SarsCov2_expected_lineage_proportions.csv" 

mtub_prop_expected <- readr::read_csv("https://raw.githubusercontent.com/mikesovic/MixviR/main/mixvir_paper/Mtub_expected_lineage_proportions.csv")

mtub_prop_observed <- readr::read_csv("./mtub_validation/estimated_proportions.csv") %>%
  dplyr::rename("Scaled Proportion Mean" = "Mean_Freq") %>%
  dplyr::mutate("Type" = "MixviR Estimate") %>%
  dplyr::select(Sample, Lineage, Type, `Scaled Proportion Mean`) %>%
  dplyr::mutate("Lineage" = stringr::str_replace_all(Lineage, "L1 \\(EAI\\)", "EAI")) %>%
  dplyr::mutate("Lineage" = stringr::str_replace_all(Lineage, "L3 \\(CAS\\)", "CAS")) %>%
  dplyr::mutate("Sample" = stringr::str_replace_all(Sample, "_split.vcf", ""))

mtub_results_prop <- dplyr::bind_rows(mtub_prop_expected, mtub_prop_observed)

mtub_lin_prop_plot <- ggplot2::ggplot(mtub_results_prop, ggplot2::aes(x = Type, y = `Scaled Proportion Mean`, fill = Lineage)) +
  ggplot2::geom_bar(stat = "identity", position = "stack") +
  ggplot2::scale_fill_brewer(palette = "Paired") +
  ggplot2::facet_wrap(ggplot2::vars(Sample)) +
  ggplot2::theme(axis.title.x = ggplot2::element_blank()) +
  ggplot2::labs(y = "Proportion In Sample",
                title = "",
                subtitle = "Estimates Of Lineage Proportions") +
  ggplot2::theme(legend.title = ggplot2::element_blank(),
                 legend.text = ggplot2::element_text(size = 10),
                 strip.text = ggplot2::element_text(size = 12),
                 axis.text = ggplot2::element_text(size = 11),
                 title = ggplot2::element_text(size = 15)) +
  ggplot2::scale_x_discrete(labels= c("Expected", "MixviR\nEstimate"))

ggplot2::ggsave(filename = "./Figure_3B.tiff", dpi = 600)
```

### Combine M. tuberculosis Plots (Fig 3)

```{r, fig.width=10}
mtub_plot <- mtub_mut_prop_plot + mtub_lin_prop_plot
mtub_plot

ggplot2::ggsave(filename = "./Figure_3.tiff", dpi = 600)
```

## Wastewater Example Plots (Fig 4)

### Call Mutations

#note the infiles used here are in csv format instead of vcf
#running the following code requires downloading the set of csv infiles available at "https://github.com/mikesovic/MixviR/tree/main/mixvir_paper/ww_infiles" and updating the *sample.dir* path in the call_mutations() function below to the directory containing the downloaded files.

```{r}
ww_exp_muts <- MixviR::call_mutations(sample.dir = "ww_example/infiles", 
               csv.infiles = TRUE, 
               name.sep = ".csv", 
               reference = "Wuhan")
```

### Lineage Present Plot (Fig 4A)

```{r, fig.width = 10}
depth.from.all.muts <- TRUE
use.mean <- TRUE
scale.freqs <- TRUE

samp_data <- ww_exp_muts %>%
      dplyr::select(SAMP_NAME, CHR, POS, GENE, ALT_ID, AF, DP)

#merge in dates
dates_df <- readr::read_csv("https://raw.githubusercontent.com/mikesovic/MixviR/main/mixvir_paper/ww_example_location_date.csv", col_types = "cci")

samp_data <- dplyr::left_join(x = samp_data, y = dates_df, by = "SAMP_NAME") %>%
      dplyr::mutate("date" = lubridate::mdy(DATE)) %>%
      dplyr::mutate("Location" = as.character(LOCATION)) %>%
      dplyr::select(-DATE, -LOCATION)

lineage_muts <- readr::read_csv("https://raw.githubusercontent.com/mikesovic/MixviR/main/mutation_files/outbreak_20220217.csv",
          col_types = readr::cols_only(Gene = readr::col_character(),
                                       Mutation = readr::col_character(),
                                       Lineage = readr::col_character())) %>%
        tidyr::unite("ALT_ID",
                     Gene, Mutation,
                     sep = "_") 
      
#identify mutations that are shared by more than one lineage
duplicated <- lineage_muts$ALT_ID[duplicated(lineage_muts$ALT_ID)]

#flag mutations that are shared by more than one lineage
lineage_muts <- lineage_muts %>%
  dplyr::mutate("characteristic" = ifelse(!ALT_ID %in% duplicated, "Y", "N"))
      
samp_data <- samp_data %>%
     dplyr::filter(Location == "Site 1") 
      
#create master df that will store data for all selected lineages/variants
all_summary <- data.frame()
    
Lineages <- c("Alpha", "Delta", "Omicron_BA.1")            
for (i in 1:length(Lineages)){  #loop over each selected lineage
  #get the mutations characteristic of the current lineage
  muts <- lineage_muts %>%
    dplyr::filter(characteristic == "Y") %>%
    dplyr::filter(Lineage %in% Lineages[[i]]) %>%
    dplyr::pull(ALT_ID)
  
  #get the number of mutations characteristic of the current lineage
  lineage_n <- length(muts)
        
#get proportion of characteristic mutations observed on each date for current Location
summary <- samp_data %>%
      #remove any duplicated mutations, keeping one with highest read depth, and summarize
      dplyr::group_by(SAMP_NAME, date, ALT_ID) %>%
      dplyr::arrange(desc(DP), .by_group = TRUE) %>%
      dplyr::ungroup() %>%
      dplyr::distinct(SAMP_NAME, date, ALT_ID, .keep_all = TRUE) %>%
      dplyr::group_by(SAMP_NAME, date) %>%
      dplyr::summarise("Proportion Present" = sum(ALT_ID[AF > 0 & DP >= 0] %in% muts)/lineage_n,
                "Lineage" = paste0(Lineages[[i]], ""),
                "Total Lineage-Characteristic Muts" = lineage_n,
                "Mean_Coverage_All" = mean(DP[DP >= 0], na.rm = TRUE),
                "DP_n_all" = sum(!is.na(DP[DP >= 0])),
                "PropHighDepth" = sum(DP[ALT_ID %in% muts] >= 0)/lineage_n) %>%
      dplyr::ungroup() %>% 
      dplyr::mutate("ID" = paste0(SAMP_NAME, "_", date))
  
#recalculate Mean Coverage based on target mutations only and add in to 'summary' from above
summary2 <- samp_data %>%
  dplyr::filter(ALT_ID %in% muts) %>%
  dplyr::group_by(SAMP_NAME, date, ALT_ID) %>%
  dplyr::arrange(desc(DP), .by_group = TRUE) %>%
  dplyr::ungroup() %>%
  dplyr::distinct(SAMP_NAME, date, ALT_ID, .keep_all = TRUE) %>%
  dplyr::group_by(SAMP_NAME, date) %>%
  dplyr::summarise("Mean Coverage" = mean(DP[DP >= 0]),
                   "DP_n" = sum(!is.na(DP[DP >= 0]))) %>%
  dplyr::ungroup() %>%
  dplyr::mutate("ID" = paste0(SAMP_NAME, "_", date)) %>%
  dplyr::select(ID, `Mean Coverage`, DP_n)

summary <- dplyr::left_join(x = summary, y = summary2, by = "ID")

#add data for current lineage to master df
all_summary <- dplyr::bind_rows(all_summary, summary)

} 
            
#use days since earliest sample date as x axis (lubridate interval)
ints <- lubridate::interval(start = min(all_summary$date), end = all_summary$date)

#convert seconds to days
all_summary <- all_summary %>%
  dplyr::mutate("days" = lubridate::int_length(ints)/86400)

#get vectors for labeling x-axis
break_dates <- unique(all_summary$date)
break_days <- unique(all_summary$days)

#check which type of mean depth to plot
if (depth.from.all.muts == TRUE) {
  
  #replace Mean Coverage col with data from Mean_Coverage_All
  all_summary <- all_summary %>% 
    dplyr::select(-`Mean Coverage`) %>%
    dplyr::rename("Mean Coverage" = "Mean_Coverage_All") %>%
    dplyr::select(-DP_n) %>%
    dplyr::rename("DP_n" = "DP_n_all")
}
            
#create new column for fill to represent seq depth. Need a constant range for the scale for this to be useful, so capping it at 1000.
all_summary <- all_summary %>%
  dplyr::mutate("mean_cov_scaled" = `Mean Coverage`) %>%
  dplyr::mutate_at(dplyr::vars(mean_cov_scaled), ~replace(., mean_cov_scaled > 1000, 1000)) %>%
  dplyr::mutate("Proportion Present" = round(`Proportion Present`, 3),
                "PropHighDepth" = round(`PropHighDepth`, 3),
                "Mean Coverage" = round(`Mean Coverage`, 3))

            
#generate plot
pres_plot <- all_summary %>%
    ggplot2::ggplot(ggplot2::aes(x = days,
                                 y = `Proportion Present`)) +
    ggplot2::geom_line(ggplot2::aes(colour = Lineage, group = Lineage), size = 1.5) +
    ggplot2::coord_cartesian(ylim = c(0,1)) +
    ggplot2::theme_classic() +
    ggplot2::scale_color_brewer(palette = "Paired") +
    ggplot2::ggtitle(paste0("Proportion of Lineage-Characteristic Mutations Present: ", "Site 1")) +
    ggplot2::scale_x_continuous(breaks = break_days, labels = break_dates) +
    ggplot2::theme(axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(angle = 45, size = 10, vjust = 0.4),
                   axis.text.y = ggplot2::element_text(size = 10),
                   plot.title = ggplot2::element_text (color = "black", size= 13, face="bold"),
                   axis.title.y = ggplot2::element_blank(),
                   legend.text = ggplot2::element_text(size = 11),
                   legend.title = ggplot2::element_blank()) +
    ggplot2::geom_hline(yintercept = 0.5, color = "red", linetype = 2, alpha = 0.6) +
    ggplot2::geom_point(data = all_summary, mapping = ggplot2::aes(x = days,
                                                                   y = `Proportion Present`,
                                                                   fill = mean_cov_scaled), size = 2.5, stroke = 0, show.legend = FALSE) +
    ggplot2::scale_fill_gradient(low = "#56B1F7",
                                  high = "#132B43",
                                  limits = c(0,1000))
     

pres_plot

ggplot2::ggsave(filename = "./Figure_4A.tiff", dpi = 600)                            
```

### Frequency Plot (Fig 4B)

```{r, fig.width=10}
samp_data <- ww_exp_muts %>%
      dplyr::select(SAMP_NAME, CHR, POS, GENE, ALT_ID, AF, DP)

#merge in dates
dates_df <- readr::read_csv("https://raw.githubusercontent.com/mikesovic/MixviR/main/mixvir_paper/ww_example_location_date.csv", col_types = "cci")

samp_data <- dplyr::left_join(x = samp_data, y = dates_df, by = "SAMP_NAME") %>%
      dplyr::mutate("date" = lubridate::mdy(DATE)) %>%
      dplyr::mutate("Location" = as.character(LOCATION)) %>%
      dplyr::select(-DATE, -LOCATION)

lineage_muts <- readr::read_csv("https://raw.githubusercontent.com/mikesovic/MixviR/main/mutation_files/outbreak_20220217.csv",
          col_types = readr::cols_only(Gene = readr::col_character(),
                                       Mutation = readr::col_character(),
                                       Lineage = readr::col_character())) %>%
        tidyr::unite("ALT_ID",
                     Gene, Mutation,
                     sep = "_") 
      
#identify mutations that are shared by more than one lineage
duplicated <- lineage_muts$ALT_ID[duplicated(lineage_muts$ALT_ID)]

#flag mutations that are shared by more than one lineage
lineage_muts <- lineage_muts %>%
  dplyr::mutate("characteristic" = ifelse(!ALT_ID %in% duplicated, "Y", "N"))
      
samp_data <- samp_data %>%
     dplyr::filter(Location == 'Site 1') 

all_summary <- data.frame()

for (i in 1:length(Lineages)){
 muts <- lineage_muts %>%
                dplyr::filter(characteristic == "Y") %>%
                dplyr::filter(Lineage %in% Lineages[[i]]) %>%
                dplyr::select(ALT_ID) %>%
                unlist()
              
              #get the number of mutations characteristic of the current lineage
              lineage_n <- length(muts)

  summary <- samp_data %>%
                  dplyr::filter(ALT_ID %in% muts) %>%
                  dplyr::group_by(SAMP_NAME, date, ALT_ID) %>%
                  dplyr::arrange(desc(DP), .by_group = TRUE) %>%
                  dplyr::ungroup() %>%
                  dplyr::distinct(SAMP_NAME, date, ALT_ID, .keep_all = TRUE) %>%
                  dplyr::group_by(SAMP_NAME, date) %>%
                  dplyr::summarise("Proportion Present" = length(ALT_ID[AF > 0 & DP >= 0])/lineage_n,
                   "Lineage" = paste0(Lineages[[i]], ""),
                   "Estimated Freq" = mean(AF[AF > 0 & DP >= 0]),
                   "muts_present" = paste0(c(sort(unique(ALT_ID[AF > 0 & DP >= 0]))), collapse = ";"),
                   "muts_absent" = paste0(c(sort(unique(muts[!muts %in% ALT_ID[AF > 0 & DP >= 0]]))), collapse = ";"),
                   "lab1" = "\nMuts Present: ",
                   "lab2" = "\n\nMuts Absent: ") %>%
  dplyr::mutate("muts_list" = glue::glue("{lab1}{muts_present}{lab2}{muts_absent}"),
                "muts_list" = stringr::str_replace_all(string = muts_list, pattern = "NA\n", replacement = ""),
                "muts_list" = gsub('(.{1,30})(;|$)', '\\1;\n', muts_list)) %>%
  dplyr::mutate_at(dplyr::vars(`Estimated Freq`), ~replace(., is.nan(.), 0)) %>%
  dplyr::mutate_at(dplyr::vars(`Estimated Freq`), ~replace(., `Proportion Present` < 0.5, 0))
  
 #add data for current lineage to master df
 all_summary <- dplyr::bind_rows(all_summary, summary) 
}

#use days since earliest sample date as x axis (lubridate interval)
ints <- lubridate::interval(start = min(all_summary$date), end = all_summary$date)

#convert seconds to days
all_summary <- all_summary %>%
  dplyr::mutate("days" = lubridate::int_length(ints)/86400)

#get vectors for labeling x-axis
break_dates <- unique(all_summary$date)
break_days <- unique(all_summary$days)

#generate plot
#only plot bars for lineages defined as "present" in the sample based on chosen proportion threshold 
 
if (scale.freqs == TRUE) {
  all_summary_sum <- all_summary %>% 
    dplyr::ungroup() %>%
    dplyr::group_by(date) %>%
    dplyr::summarise("sum" = sum(`Estimated Freq`)) %>%
    dplyr::mutate("over1" = ifelse(sum > 1, "Y", "N"))
  
  all_summary_sum <- dplyr::left_join(x = all_summary, y = all_summary_sum, by = "date")
   
  to_scale <- all_summary_sum %>%
    dplyr::filter(over1 == "Y") %>% 
    dplyr::group_by(date) %>% 
    dplyr::mutate("Estimated Freq" = `Estimated Freq`/sum(`Estimated Freq`))
    
  unscaled <- all_summary_sum %>%
    dplyr::filter(over1 == "N")
  
  all_summary <- dplyr::bind_rows(to_scale, unscaled)
}

if (use.mean == TRUE) {
  measure <- "Mean"
} else {
  measure <- "Median"
}

all_summary <- all_summary %>%
  dplyr::mutate("Proportion Present" = round(`Proportion Present`, 3),
                "Estimated Freq" = round(`Estimated Freq`,3))


prop_plot <- all_summary %>%
              dplyr::filter(`Estimated Freq` > 0) %>%
              ggplot2::ggplot(ggplot2::aes(x = days, y = `Estimated Freq`, fill = Lineage)) +
              ggplot2::geom_bar(stat = "identity",
                       position = "stack") +
              ggplot2::theme_classic() +
              ggplot2::scale_fill_brewer(palette = "Paired") +
              ggplot2::ggtitle(paste0("Estimated Frequency of Each Lineage Present: ", "Site 1")) +
              ggplot2::scale_x_continuous(breaks = break_days, labels = break_dates) +
              ggplot2::theme(axis.title.x = ggplot2::element_blank(),
                    axis.text.x = ggplot2::element_text(angle = 45, vjust = 0.4),
                    axis.text = ggplot2::element_text(size = 10),
                    plot.title = ggplot2::element_text (color = "black", size= 11, face="bold"),
                    axis.title.y = ggplot2::element_blank(),
                    legend.text = ggplot2::element_text(size = 11),
                    legend.title = ggplot2::element_blank()) +
              ggplot2::labs(y = paste0("Estimated Frequency (", measure, ")"))

prop_plot

ggplot2::ggsave(filename = "./Figure_4B.tiff", dpi = 600)  
```


# Supplemental Analyses

## Regression Plot (Fig S1)

```{r}
mtub_data <- mtub_results_prop %>%
  dplyr::select(Sample, Lineage, Type, `Scaled Proportion Mean`) %>%
  tidyr::pivot_wider(names_from = Type, values_from = `Scaled Proportion Mean`)

mtub_rsquared <- summary(lm(mtub_data$`MixviR Estimate`~mtub_data$Expected))[[9]]

lab_df <- data.frame("x" = c(0.3, 0.96), "y" = c(0.87, 0.02), "text" = c(paste("r^2 = ", round(mtub_rsquared, 3)), "B"))

mtub_plot <- mtub_data %>% ggplot2::ggplot(ggplot2::aes(x = Expected, y = `MixviR Estimate`, shape = Lineage)) +
  ggplot2::geom_point(size = 2.5) +
  ggplot2::geom_smooth(method = "lm", inherit.aes = FALSE, 
              mapping = ggplot2::aes(x = Expected, y = `MixviR Estimate`), color = "black") +
    ggplot2::theme_classic() +
    ggplot2::labs(title = "", 
        subtitle = "M. tuberculosis",
         x = "Expected Proportion",
        y = ggplot2::element_blank()) +
    ggplot2::theme(legend.position = c(0.80, 0.16), 
      legend.key.size = ggplot2::unit(0.5, 'cm'),
      legend.text = ggplot2::element_text(size = 12),
      strip.text = ggplot2::element_text(size = 12),
      axis.text = ggplot2::element_text(size = 11),
      title = ggplot2::element_text(size = 15)) +
  ggplot2::coord_cartesian(xlim = c(0,1)) +
  ggplot2::geom_label(data = lab_df %>% dplyr::slice(1), mapping = ggplot2::aes(x = x, y = y, label = text), inherit.aes = FALSE, size = 6) +
  ggplot2::geom_text(data = lab_df %>% dplyr::slice(2), mapping = ggplot2::aes(x = x, y = y, label = text), inherit.aes = FALSE, size = 11)

#Get Covid plot
sars_data <- covid_results_prop %>%
  dplyr::select(Sample, Lineage, Type, `Scaled Proportion Mean`) %>%
  tidyr::pivot_wider(names_from = Type, values_from = `Scaled Proportion Mean`)

sars_rsquared <- summary(lm(sars_data$`MixviR Estimate`~sars_data$Expected))[[9]]

lab_df <- data.frame("x" = c(0.3, 0.96), "y" = c(0.9, 0.02), "text" = c(paste("r^2 = ", round(sars_rsquared, 3)), "A"))

sars_plot <- sars_data %>% ggplot2::ggplot(ggplot2::aes(x = Expected, y = `MixviR Estimate`, shape = Lineage)) +
  ggplot2::geom_point(size = 2.5) +
  ggplot2::geom_smooth(method = "lm", inherit.aes = FALSE, 
              mapping = ggplot2::aes(x = Expected, y = `MixviR Estimate`), color = "black") +
  ggplot2::theme_classic() +
  ggplot2::labs(title = "Mixvir Estimates Of Lineage Frequencies In Mixed Samples",
    subtitle = "SARS-CoV-2",
    x = "Expected Proportion") +
  ggplot2::theme(legend.position = c(0.80, 0.18), 
                 legend.key.size = ggplot2::unit(0.5, 'cm'),
                 legend.text = ggplot2::element_text(size = 12),
                 strip.text = ggplot2::element_text(size = 12),
                 axis.text = ggplot2::element_text(size = 11),
                 title = ggplot2::element_text(size = 13.5)) +
  ggplot2::coord_cartesian(ylim = c(0,1)) +
  ggplot2::geom_label(data = lab_df %>% dplyr::slice(1), mapping = ggplot2::aes(x = x, y = y, label = text), inherit.aes = FALSE, size = 6) +
  ggplot2::geom_text(data = lab_df %>% dplyr::slice(2), mapping = ggplot2::aes(x = x, y = y, label = text), inherit.aes = FALSE, size = 11)

plotS1 <- sars_plot + mtub_plot
plotS1

ggplot2::ggsave(file = "./Figure_S1.tiff", dpi = 600)
```

## Shiny Examples (Figs S2-S5)

Examples included were screenshots from tabs in Shiny dashboard.

```{r, eval = FALSE}
if (interactive) {
MixviR::explore_mutations(muts.df = ww_exp_muts,
                          dates = "https://raw.githubusercontent.com/mikesovic/MixviR/main/mixvir_paper/ww_example_location_date.csv",
                  lineage.muts = "https://raw.githubusercontent.com/mikesovic/MixviR/main/mutation_files/outbreak_20220217.csv")
}

```
